### PR TITLE
ENH: Make regtap service aware

### PR DESCRIPTION
**Pull request by msdemlei**
This PR contains a set of changes addressing the fixable but not yet
fixed issues raised by Tom in
https://wiki.ivoa.net/internal/IVOA/InterOpOct2022Reg/PyVOandNAVORegTAP.pdf

PyVO now properly checks whether the server can do something not in
RegTAP 1.0 before trying it.  This concerns the optimisation of the
Freetext constraint as well as the STC constraints.

To make this reasonably easy to write, this retrofits some methods to
non-registry code:

* get_tap_cap to TAPService
* get_adql to tr.TableAccesss
* get_feature_list, get_feature, and get_udf to tr.Language

### CODE REVIEW

File: docs/registry/index.rst
Comment (bsipocz) [inline docs/registry/index.rst:351]: I know that you don't yet want review, so I leave only one comment:

we should be able to do this with a more programatic way, from inside the python interpreter.
Whether to use the astropy config system, or also provide ways to override the defaults in a session, or be able to do both is a detail, but I very strongly think it should be doable without shell variables.

**Comment by msdemlei** [comment]
On Wed, Dec 21, 2022 at 04:31:50PM -0800, Brigitta Sipőcz wrote:
> we should be able to do this with a more programatic way, from
> inside the python interpreter.

You can (see registry.tests.test_regtap.test_endpoint_switching; it's
regtap.choose_RegTAP_service).  But the env var method is IMHO still
important in the medium run -- I am hoping to set up RegTAP endpoints
in less well-connected regions, and then people there can
pre-configure their desired endpoints, which does make a noticable
difference (I have first-hand experience on this from South America
and China).

By the way, I'm also trying to talk Mark Taylor into interpreting
that environment variable.

**Comment by bsipocz** [comment]
I milestoned this for 1.5, with the mindset that this is rather an enhancement rather than a bugfix.

**Comment by bsipocz** [comment]
I think this is the only thing that holds up 1.5, if we want to maybe do a release soonish? I recall using this branch when testing the SIA2 registry stuff, so overall not sure how much more is missing for this to be ready to be merged. Maybe it's only a rebase from Markus and a review from Tom?

**Comment by bsipocz** [comment]
(I'm mostly pushing for the release as it would be nice to get that SIA2 registry search compatibility out of the door)

**Comment by msdemlei** [comment]
On Mon, May 08, 2023 at 07:10:06AM -0700, Brigitta Sipőcz wrote:
> I think this is the only thing that holds up 1.5, if we want to
> maybe do a release soonish? I recall using this branch when testing
> the SIA2 registry stuff, so overall not sure how much more is
> missing for this to be ready to be merged. Maybe it's only a rebase
> from Markus and a review from Tom?

I'd love to see this in RSN -- it's real progress towards fulfilling
the promise of interoperating publishing registries.  I've hence done
the rebase (oh boy, never rebase over a whitespace-everywhere commit)
and I hope I've not messed things up too much.  Tom, a review would
be most appreciated.  While we're all here, I'd actually be happy to
do that in presence.

I expect the test suite will throw two errors: One is about a QTable
(I'll investigate tomorrow -- alternatively, advice appreciated), the
other is a failing query on the StScI service.  I *think* that's just
a matter of rolling out a fix on their side.  Tom?

**Comment by bsipocz** [comment]
> I've hence done
the rebase (oh boy, never rebase over a whitespace-everywhere commit)
and I hope I've not messed things up too much. 

Could you push this back to the branch here?

**Comment by bsipocz** [comment]
I made some CI improvements, so will go ahead and rebase again and flip the switch to move this out from the draft PR status.

**Comment by msdemlei** [comment]
On Wed, May 10, 2023 at 01:13:24AM -0700, Brigitta Sipőcz wrote:
> @msdemlei - I wonder why CI doesn't see those failures you mention before, are they covered by tests or it's about use cases you haven't covered in the test suite yet?

So much the better if the CI doesn't see them.  Note that I'm running
bleeding-edge pyvo against astropy and all the pytest packages as in
Debian stable, so it's totally possible we're just seeing legacy
behaviour blowing up.

[Side rant: Given today's standards of software engineering, I'm
actually rather impressed I can develop pyvo in that configuration.
You see, I can't for astropy, which needs a bleeding-edge venv (which
I'm unhappy about given the perils of pulling executable code from
only very lightly curated repos).  Thanks to everyone for keeping
pyvo Debian stable-friedly!]

**Comment by bsipocz** [comment]
>  bleeding-edge venv

I would certainly raise this issue upstream. IMO it should all work in a stable environment and when it doesn't it certainly should be looked at as a bug rather than a feature. Getting the latest and greatest cool tools shouldn't be a requirement to do dev work.

**Comment by msdemlei** [comment]
Um... I just noticed this still lingers in the doldrums where I'd say it's a big improvement over what we have in terms of letting people switch registries.

So... @tomdonaldson – could you spare an eye?  Or perhaps @theresadower?

**Comment by bsipocz** [comment]
>  Again unrelated to this, and it'd be lovely if someone could fix it 

I'm afraid this might have been rebased on an outdated branch. I started to look into a rebase on `main`, and some of the test failures are gone, but I see new ones. I expect some more cleaning would fix the unrelated ones of those. I'm happy to push back to this branch (but that won't be before ~middle/late next week)), or have you do another rebase yourself?

File: docs/registry/index.rst
Comment (bsipocz) [inline docs/registry/index.rst:549]: This causes one of the test failures

```suggestion
  http://voparis-rr.obspm.fr/tap
```

**Comment by bsipocz** [comment]
I went ahead and pushed the rebased version back to this branch as I managed to clear up the test failures that were due to the conflicts. I still see the endpoint switching issue @msdemlei mentioned above as well as an unrelated looking `TestXSIType.test_bad_type` (I'm about to open a separate cleanup PR, if I see this one in there, I'll try to fix it there).
And there is a doctest one that I'll comment/suggest in the review I'm about to submit.

So, bottom line, would be nice to fix that one related test and make this work and merge and then finally release v1.5.

**Review by bsipocz — COMMENTED** [review]
A lot of nitpicky comments with suggestions that would either help with fixing the links in the docsbuild (they could be part of https://github.com/astropy/pyvo/pull/416, but I feel it's better to get things fixed when they pop-up), or suggestions for the failing doctest.

As none of these are critical (except fixing the failing doctest), that take them or leave them, they should not block merging this PR. 

Also, could you please add a changelog entry?

**Comment by bsipocz** [comment]
I went ahead and fixed the `TestXSIType.test_bad_type` here, as while it was unrelated, it got triggered due to not using pytest properly for warning testing (and a newly issued, but eventually ignored socket warning got popped up and asserted in the test as opposed to the expected one)

File: docs/registry/index.rst
Comment (msdemlei) [reply docs/registry/index.rst:546]: Mhhh... I actually prefer broken lines and the loop expression on an extra line...

File: pyvo/dal/tap.py
Comment (msdemlei) [reply pyvo/dal/tap.py:130]: Absolutely, yes.

File: pyvo/registry/rtcons.py
Comment (msdemlei) [reply pyvo/registry/rtcons.py:61]: Yes... this is what happens when you change your naming preferences in between edits.  Fixed.

File: pyvo/registry/regtap.py
Comment (msdemlei) [reply pyvo/registry/regtap.py:108]: Please hand me a brown bag.  Fixed in the next commit as well.

File: docs/registry/index.rst
Comment (msdemlei) [reply docs/registry/index.rst:549]: Thanks...

**Comment by msdemlei** [comment]
So... I hope I've addressed most of @bsipocz's concerns; I can go hunting a few more of the bad in-documentation references once I'm sure what the right pattern is.

We still have test_endpoint_switching fail, and I think we ought to get it fixed before merging this (it's serious, and it's the whole point of this PR) -- @tomdonaldson, can you advise whether we should put in some workaround (which?) or whether you can fix the service at your end?

Also, I'm a bit at the limits of my git-fu – after the maintainer edits, the branch this PR is fed from is now astropy/m-t-s-a rather than msdemlei/m-t-s-a, right?  That's by design?

**Comment by bsipocz** [comment]
>  I can go hunting a few more of the

I would say, don't waste your time on it, as they are difficult to spot when the ci doesn't yet pick them up. 
It was really just a heads up, we'll fix the ones I found and leave the rest for the other pr.

File: pyvo/dal/tap.py
Comment (bsipocz) [reply pyvo/dal/tap.py]: Wrapping up #416 and putting it into CI. I was doing a little bit of work on that PR, but the main issue is that we need to make some decisions about what's in the public API (and thus documented) and what's not. Currently I went ahead and ignored a couple of classes that we said is not for public consumption, yet we try to link to them as e.g. they are used as base classes.

Comment (bsipocz) [reply pyvo/dal/tap.py]: (btw, wrapping up #416 is a bit of a hustle as there is an underlying sphinx-osx issue that in theory has a workaround but I never managed to make it work here, so I have to rely on CI for the iterations)

**Comment by bsipocz** [comment]
> We still have test_endpoint_switching fail, and I think we ought to get it fixed before merging this (it's serious, and it's the whole point of this PR)

Do we have a third registry we could try this with? I mean, yes, the whole point of this is to make sure it works with the ST one, but if that one is expected to be a bit flaky we may want to test with a third option?

**Comment by tomdonaldson** [comment]
> We still have test_endpoint_switching fail, and I think we ought to get it fixed before merging this (it's serious, and it's the whole point of this PR) -- @tomdonaldson, can you advise whether we should put in some workaround (which?) or whether you can fix the service at your end?
> 
@msdemlei I will see what's up with this within the next day.

**Comment by msdemlei** [comment]
On Tue, Dec 05, 2023 at 09:09:34AM -0800, Brigitta Sipőcz wrote:
> > We still have test_endpoint_switching fail, and I think we ought to get it fixed before merging this (it's serious, and it's the whole point of this PR)
>
> Do we have a third registry we could try this with? I mean, yes,
> the whole point of this is to make sure it works with the ST one,
> but if that one is expected to be a bit flaky we may want to test
> with a third option?

There is the ESAC registry, but that has gone somewhat into
disrepair (not only for us).  There is work being done to fix this
(and, most importantly for me, finally do a second implementation of
RegTAP 1.2), but... oh, I just realise this is updated.  Hey!  Great!
The endpoint is https://registry.euro-vo.org/regtap/tap.

I have just tried it, but there's still a parser problem at their
end.  I'll contact them.  Still, I think I'd be too cheap to add
remote-data tests for all the various endpoints, and for STScI I
think there's really not much missing, just a name resolution bug
that (from the outside) would seem reasonably simple to fix.

**Comment by bsipocz** [comment]
FYI, with the fix in #503, I still see an issue coming from the doctest and it looks relevant:

```
================================================================================= FAILURES =================================================================================
___________________________________________________________________________ [doctest] index.rst ____________________________________________________________________________
145                     Asiago Supernova Catalogue (Version 2008-Mar) ... conesearch, tap#aux, web
146        Sloan Digital Sky Survey-II Supernova Survey (Sako+, 2018) ... conesearch, tap#aux, web
147   ...
148 
149 And to look for tap resources *in* a specific cone, you would do
150 
151 .. doctest-remote-data::
152 
153   >>> from astropy.coordinates import SkyCoord
154   >>> registry.search(registry.Servicetype("tap"),
UNEXPECTED EXCEPTION: RegTAPFeatureMissing: stc_spatial missing on current RegTAP service
Traceback (most recent call last):
  File "/Users/bsipocz/.pyenv/versions/3.11.0/lib/python3.11/doctest.py", line 1350, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest index.rst[9]>", line 1, in <module>
  File "/Users/bsipocz/munka/devel/pyvo/pyvo/registry/regtap.py", line 247, in search
    get_RegTAP_query(*constraints,
  File "/Users/bsipocz/munka/devel/pyvo/pyvo/registry/regtap.py", line 160, in get_RegTAP_query
    return rtcons.build_regtap_query(constraints, service)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bsipocz/munka/devel/pyvo/pyvo/registry/rtcons.py", line 906, in build_regtap_query
    "(" + constraint.get_search_condition(service) + ")")
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bsipocz/munka/devel/pyvo/pyvo/registry/rtcons.py", line 691, in get_search_condition
    raise RegTAPFeatureMissing("stc_spatial missing on current RegTAP service")
pyvo.registry.rtcons.RegTAPFeatureMissing: stc_spatial missing on current RegTAP service
/Users/bsipocz/munka/devel/pyvo/docs/registry/index.rst:154: UnexpectedException
========================================================================= short test summary info ==========================================================================
FAILED docs/registry/index.rst::index.rst
================================================================= 1 failed, 2 passed in 245.14s (0:04:05) ==================================================================
```

**Comment by bsipocz** [comment]
btw, this failing doctest shows that it would be in fact beneficial to include more non-mocked tests that reaches the services. Or of course we can indeed think of the docs as a way of increasing the test coverage, we just need to make sure it's sufficiently goes into the corner cases.

**Comment by tomdonaldson** [comment]
Looking at the test failure in https://github.com/astropy/pyvo/blob/fe3f909d10da8c6c560afe9e682a4ce07c8be942/pyvo/registry/tests/test_regtap.py#L890
MAST is giving the following error for the regtap query generated by pyvo:
```
Exception translating ADQL: 1 unresolved identifiers: res_subject [l.15 c.84 - l.15 c.110]!
  - Unknown column "rr.res_subject.res_subject" !
```
The ADQL submitted for that query is shown as `or_query` in the code block below.

The `WHERE` clause in that query is generated by pyvo as a workaround for RegTAP services that don't support `UNION` and uses `OR`s instead.  

It looks to me like the `OR` version of the query may be flawed since the GAVO RegTAP service issues a similar complaint about it.  I suppose this is because `rr.res_subject` isn't really part of the rest of the query, but that's just a vague notion I don't have a quick suggestion to address it.

See the example code below for how both the `OR` and `UNION` versions of the query behave for MAST and GAVO.

```python
import pyvo as vo


gavo_regtap_service = vo.dal.TAPService("http://reg.g-vo.org/tap")
mast_regtap_service = vo.dal.TAPService("https://vao.stsci.edu/RegTAP/TapService.aspx")

# The default regtap query uses UNION in the WHERE clause.
union_query = """
SELECT
ivoid, res_type, short_name, res_title, content_level, res_description, reference_url, creator_seq, created, updated, rights, content_type, source_format, source_value, region_of_regard, waveband,
  ivo_string_agg(COALESCE(access_url, ''), ':::py VO sep:::') AS access_urls,
  ivo_string_agg(COALESCE(standard_id, ''), ':::py VO sep:::') AS standard_ids,
  ivo_string_agg(COALESCE(intf_type, ''), ':::py VO sep:::') AS intf_types,
  ivo_string_agg(COALESCE(intf_role, ''), ':::py VO sep:::') AS intf_roles, alt_identifier
FROM
rr.resource
NATURAL LEFT OUTER JOIN rr.capability
NATURAL LEFT OUTER JOIN rr.interface
NATURAL LEFT OUTER JOIN rr.alt_identifier
WHERE
(ivoid IN (SELECT ivoid FROM rr.resource WHERE 1=ivo_hasword(res_description, 'wirr') UNION SELECT ivoid FROM rr.resource WHERE 1=ivo_hasword(res_title, 'wirr') UNION SELECT ivoid FROM rr.res_subject WHERE rr.res_subject.res_subject ILIKE '%wirr%'))
GROUP BY
ivoid, res_type, short_name, res_title, content_level, res_description, reference_url, creator_seq, created, updated, rights, content_type, source_format, source_value, region_of_regard, waveband, alt_identifier"""


# The or query is intended for regtap services that don't support UNION.
or_query = """
SELECT
ivoid, res_type, short_name, res_title, content_level, res_description, reference_url, creator_seq,
created, updated, rights, content_type, source_format, source_value, region_of_regard, waveband,
  ivo_string_agg(COALESCE(access_url, ''), ':::py VO sep:::') AS access_urls,
  ivo_string_agg(COALESCE(standard_id, ''), ':::py VO sep:::') AS standard_ids,
  ivo_string_agg(COALESCE(intf_type, ''), ':::py VO sep:::') AS intf_types,
  ivo_string_agg(COALESCE(intf_role, ''), ':::py VO sep:::') AS intf_roles, alt_identifier
FROM
rr.resource
NATURAL LEFT OUTER JOIN rr.capability
NATURAL LEFT OUTER JOIN rr.interface
NATURAL LEFT OUTER JOIN rr.alt_identifier
WHERE
(( 1=ivo_hasword(res_description, 'wirr') OR  1=ivo_hasword(res_title, 'wirr') OR  rr.res_subject.res_subject ILIKE '%wirr%'))
GROUP BY
ivoid, res_type, short_name, res_title, content_level, res_description, reference_url, creator_seq, created, updated, rights,
content_type, source_format, source_value, region_of_regard, waveband, alt_identifier"""


def regtap_adql_query(service, adql):
    result = service.search(adql)
    print(result)


if __name__ == "__main__":
    try:
        print("\nTrying MAST RegTAP with OR query...")
        regtap_adql_query(mast_regtap_service, or_query)
    except Exception as e:
        print(e)

    try:
        print("\nTrying MAST RegTAP with UNION query (unsupported)...")
        regtap_adql_query(mast_regtap_service, union_query)
    except Exception as e:
        print(e)

    try:
        print("\nTrying GAVO RegTAP with OR query...")
        regtap_adql_query(gavo_regtap_service, or_query)
    except Exception as e:
        print(e)

    try:
        print("\nTrying GAVO RegTAP with UNION query...")
        regtap_adql_query(gavo_regtap_service, union_query)
    except Exception as e:
        print(e)
```

**Comment by bsipocz** [comment]
gentle ping that once this one gets wrapped up we can do the long-awaited 1.5 release 😄

**Comment by msdemlei** [comment]
On Wed, Dec 13, 2023 at 08:52:28AM -0800, Brigitta Sipőcz wrote:
> gentle ping that once this one gets wrapped up we can do the
> long-awaited 1.5 release 😄

Oh, I'm sorry.  If forgot to push my latest changes.  And I hadn't
put in my changelog entry, either, as I just noticed.  I'll pay more
attention to the CI this time, promised.

Thanks for your patience!

**Comment by msdemlei** [comment]
Ok, and I've rebased to current main.  @tomdonaldson, can you perhaps add your placet?

**Review by tomdonaldson — APPROVED** [review]
Thanks @msdemlei !  This all look good to me now.  The `_extra_tables` mechanism seems like a nice way to address the need for the extra join.

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:50]: As this is user facing, I would think the information that now it's possible to use other registries is important to list, or to list as well. 

That being said, I would go ahead and merge this as is so I can go ahead and do more tests and cleanups, but please do advise a rephrasing and I'll fix up the changelog before the release.

**Comment by bsipocz** [comment]
Thank you so much @msdemlei!

**Comment by msdemlei** [comment]
On Fri, Dec 15, 2023 at 11:37:31AM -0800, Brigitta Sipőcz wrote:
> @bsipocz commented on this pull request.
>
>
>
> > +- registry.search now introspects the TAP service's capabilities and
> +  only offers extended functionality or optimisations if the required
> +  features are present [#386]
>
> That being said, I would go ahead and merge this as is so I can go
> ahead and do more tests and cleanups, but please do advise a
> rephrasing and I'll fix up the changelog before the release.

Is it legal to have two Changelog items for one PR?  If so, I'd make
an extra item and say something like:

  You can now change the registry endpoint that registry.search uses
  at runtime using the choose_RegTAP_service function. [#386]

I think that's enough for a Changelog; whoever wants to do that kind
of thing will find in the docstring of the function both info on
global selection using IVOA_REGISTRY and how you can find alternative
endpoints.

**Comment by bsipocz** [comment]
> Is it legal to have two Changelog items for one PR?

Of course it is, thanks for the extra text!

**Pull request merged**
